DQL-204: Enhancements to the data request functionality

MarkCalvert opened the pull request:
[DQL2-32]
- Added new admin config 'datarequest_circumstances'
[DQL2-33]
Added new columns to datarequest table:
- close_circumstance
- approx_publishing_date
Added close_circumstance & approx_publishing_date values to action 'close_datarequest'
Show close_circumstance & approx_publishing_date values on 'additional_info.html'
[DQL2-31]
- Add template helper get_closing_circumstances
- Added closing circumstances dropdown to 'close_datarequest_form.html' with JS logic to show/hide condition elements
- Refactored '_dictize_datarequest' & '_undictize_datarequest_basic' methods to the 'db.py' model
- Added conditional logic for throuhout extension for 'closing_circumstances_enabled'

Commit:
[DQL2-32]
Added new admin config 'datarequest_circumstances'
[DQL2-33]
Added new columns to datareqest table:
- close_circumstance
- approx_publishing_date
Added close_circumstance & approx_publishing_date values to action 'close_datarequest'
Show close_circumstance & approx_publishing_date values on 'additional_info.html'
[DQL2-31]
- Add template helper get_closing_circumstances
- Added closing circumstances dropdown to 'close_datarequest_form.html' with JS logic to show/hide condition elements
- Refactored '_dictize_datarequest' & '_undictize_datarequest_basic' methods to the 'db.py' model
- Added conditional logic for throuhout extension for 'closing_circumstances_enabled'

duttonw commented on ckanext/datarequests/constants.py:
```diff
@@ -35,3 +35,4 @@
 DESCRIPTION_MAX_LENGTH = 1000
 COMMENT_MAX_LENGTH = DESCRIPTION_MAX_LENGTH
 DATAREQUESTS_PER_PAGE = 10
+CLOSE_CIRCUMSTANCE_MAX_LENGTH = 100
```
is 100char enough?

MarkCalvert replied:
Currently, for data.qld the close circumstance values are:
- Released as open data
- Open dataset already exists
- Partially released
- To be released as open data at a later date
- Data openly available elsewhere
- Not suitable for release as open data
- Requested data not available/cannot be compiled
- Requestor initiated closure
'Requested data not available/cannot be compiled' has the highest char length of 47, so the max of 100 is fine for data.qld current requirements but may not be for other CKAN sites?
I think 100 is good because anything over 100 characters in a select dropdown list is too long in my opinion.
What do you suggest as a max length? 255 is a common default

duttonw replied:
lets go with 255 just in case and as you said its a common default.

MarkCalvert replied:
Increased to 255

duttonw commented on requirements-dev.txt:
```diff
@@ -8,7 +8,7 @@ nose_parameterized==0.3.3
 lxml==4.5.0
 splinter>=0.13.0
--e git+https://github.com/qld-gov-au/ckanext-data-qld@develop#egg=ckanext-data_qld
+-e git+https://github.com/salsadigitalauorg/ckanext-data-qld@feature/Enhancements_to_the_data_request_functionality#egg=ckanext-data_qld
```
sometimes i think it would be awesome to allow these to be dynamic (group name) as well as switching branch/tag with pipeline parallel builds so that we can test against develop/master/latest tagged release etc.

MarkCalvert replied:
Yes that would be great if we could figure out how to do that, maybe we create a ticket in the backlog for this?

Commit:
[DQL2-31]
- Updated close accepted_dataset dropdown to use dataset id instead of name as its value
- Updated validate_datarequest_closing to use 'package_id_exists' to validate accepted_dataset_id

Commit:
[DQL2-40]
- Added BDD tests for datarequest circumstances
- Added CKAN config 'ckan.datarequests.enable_closing_circumstances = True' to test.ini
- Updated 'create-test-data.sh' to add new admin config value for 'ckan.datarequests.closing_circumstances'
-  Updated 'create-test-data.sh' to create test data datasets
- Added new BDD steps 'I log in and go to admin config page',  'I go to admin config page', 'I log in and create a datarequest'
- Updated 'datarequest.feature' to use new step 'I log in and create a datarequest

[DQL2-33]
- Updated validation error messages to Sentence casing
- Update additinal_info to conditionally show 'approx_publishing_date' or 'accepted_dataset' rows if the values exist

Commit:
Allow unit tests failures in circleci

duttonw commented on .env:
```diff
@@ -19,7 +19,7 @@ COMPOSE_PROJECT_NAME="ckanext-datarequests"
 
 # Flag to allow unit tests failures.
-ALLOW_UNIT_FAIL=0
+ALLOW_UNIT_FAIL=1
```
why did you allow failures, our qld-gov-au version did not have failing tests.

https://app.circleci.com/pipelines/github/qld-gov-au/ckanext-datarequests/82/workflows/db9fcac3-b1d5-4322-9e68-9e1f1f9cb0b8/jobs/79/steps
Duration 9m 41s
Finished 16 days ago
Waiting 2m 54s
Docker Medium

2 CPU / 4 GB RAM

20517e3c20d6ef
develop
William Dutton


 
==> Run Unit tests
 
..........................................................................................................................................................................................................................................................................................
 
----------------------------------------------------------------------
 
Ran 282 tests in 2.180s
 

 
OK

duttonw replied:
prior to unit test failures allowed errors were

```
======================================================================
FAIL: test_close_datarequest_1 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 737, in test_close_datarequest
    self._check_basic_response(datarequest, result, default_user, org, pkg)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 76, in _check_basic_response
    self.assertEquals(datarequest.id, response['id'])
AssertionError: 'example_uuidv4' != <MagicMock name='mock.dictize_datarequest().__getitem__()' id='139715380999760'>

======================================================================
FAIL: test_close_datarequest_2 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 737, in test_close_datarequest
    self._check_basic_response(datarequest, result, default_user, org, pkg)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 76, in _check_basic_response
    self.assertEquals(datarequest.id, response['id'])
AssertionError: 'example_uuidv4' != <MagicMock name='mock.dictize_datarequest().__getitem__()' id='139715382054608'>

======================================================================
FAIL: test_close_datarequest_3 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 737, in test_close_datarequest
    self._check_basic_response(datarequest, result, default_user, org, pkg)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 76, in _check_basic_response
    self.assertEquals(datarequest.id, response['id'])
AssertionError: 'example_uuidv4' != <MagicMock name='mock.dictize_datarequest().__getitem__()' id='139715383509328'>

======================================================================
FAIL: test_create_datarequest_valid (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/mock/mock.py", line 1330, in patched
    return func(*args, **keywargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 366, in test_create_datarequest_valid
    send_mail_mock.assert_called_once_with(set(['user_1', 'user_2']), 'new_datarequest', result)
  File "/app/ckan/default/lib/python2.7/site-packages/mock/mock.py", line 957, in assert_called_once_with
    return self.assert_called_with(*args, **kwargs)
  File "/app/ckan/default/lib/python2.7/site-packages/mock/mock.py", line 944, in assert_called_with
    six.raise_from(AssertionError(_error_message(cause)), cause)
  File "/app/ckan/default/lib/python2.7/site-packages/six.py", line 737, in raise_from
    raise value
AssertionError: expected call not found.
Expected: _send_mail(set(['user_2', 'user_1']), 'new_datarequest', <MagicMock name='mock.DataRequest().dictize_datarequest()' id='139715381427216'>)
Actual: _send_mail(set([]), 'new_datarequest', <MagicMock name='mock.DataRequest().dictize_datarequest()' id='139715381427216'>)

======================================================================
FAIL: test_delete_datarequest_0 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 666, in test_delete_datarequest
    self._check_basic_response(datarequest, result, default_user, org, pkg)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 76, in _check_basic_response
    self.assertEquals(datarequest.id, response['id'])
AssertionError: 'example_uuidv4' != <MagicMock name='mock.dictize_datarequest().__getitem__()' id='139715379412368'>

======================================================================
FAIL: test_delete_datarequest_1_org_id (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 666, in test_delete_datarequest
    self._check_basic_response(datarequest, result, default_user, org, pkg)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 76, in _check_basic_response
    self.assertEquals(datarequest.id, response['id'])
AssertionError: 'example_uuidv4' != <MagicMock name='mock.dictize_datarequest().__getitem__()' id='139715378751760'>

======================================================================
FAIL: test_delete_datarequest_2 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 666, in test_delete_datarequest
    self._check_basic_response(datarequest, result, default_user, org, pkg)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 76, in _check_basic_response
    self.assertEquals(datarequest.id, response['id'])
AssertionError: 'example_uuidv4' != <MagicMock name='mock.dictize_datarequest().__getitem__()' id='139715373472272'>

======================================================================
FAIL: test_delete_datarequest_3_org_id (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 666, in test_delete_datarequest
    self._check_basic_response(datarequest, result, default_user, org, pkg)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 76, in _check_basic_response
    self.assertEquals(datarequest.id, response['id'])
AssertionError: 'example_uuidv4' != <MagicMock name='mock.dictize_datarequest().__getitem__()' id='139715383669392'>

======================================================================
FAIL: test_list_datarequests_0 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 608, in test_list_datarequests
    self.assertEquals(expected_response['result'], response['result'])
AssertionError: Lists differ: [{'description': 'This is a ba... != [<MagicMock name='mock.dictize...

First differing element 0:
{'description': 'This is a basic description', 'organization_id': None, 'user': {'user': 3, 'id': 'user_id'}, 'id': 'example_uuidv4', 'close_time': None, 'user_id': 'example_uuidv4_user', 'open_time': '2020-06-10 18:41:45.498722', 'accepted_dataset': None, 'title': 'This is a title', 'accepted_dataset_id': None, 'followers': 3, 'closed': False, 'organization': None}
<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>

+ [<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>]
- [{'accepted_dataset': None,
-   'accepted_dataset_id': None,
-   'close_time': None,
-   'closed': False,
-   'description': 'This is a basic description',
-   'followers': 3,
-   'id': 'example_uuidv4',
-   'open_time': '2020-06-10 18:41:45.498722',
-   'organization': None,
-   'organization_id': None,
-   'title': 'This is a title',
-   'user': {'id': 'user_id', 'user': 3},
-   'user_id': 'example_uuidv4_user'}]

======================================================================
FAIL: test_list_datarequests_1 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 608, in test_list_datarequests
    self.assertEquals(expected_response['result'], response['result'])
AssertionError: Lists differ: [{'description': 'This is a ba... != [<MagicMock name='mock.dictize...

First differing element 0:
{'description': 'This is a basic description', 'organization_id': None, 'user': {'user': 3, 'id': 'user_id'}, 'id': 'example_uuidv4', 'close_time': None, 'user_id': 'example_uuidv4_user', 'open_time': '2020-06-10 18:41:45.498722', 'accepted_dataset': None, 'title': 'This is a title', 'accepted_dataset_id': None, 'followers': 3, 'closed': False, 'organization': None}
<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>

+ [<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>]
- [{'accepted_dataset': None,
-   'accepted_dataset_id': None,
-   'close_time': None,
-   'closed': False,
-   'description': 'This is a basic description',
-   'followers': 3,
-   'id': 'example_uuidv4',
-   'open_time': '2020-06-10 18:41:45.498722',
-   'organization': None,
-   'organization_id': None,
-   'title': 'This is a title',
-   'user': {'id': 'user_id', 'user': 3},
-   'user_id': 'example_uuidv4_user'}]

======================================================================
FAIL: test_list_datarequests_10 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 594, in test_list_datarequests
    self.assertEquals(organization_show.call_count - count, expected_organization_show_calls)
AssertionError: 0 != 3

======================================================================
FAIL: test_list_datarequests_11 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 594, in test_list_datarequests
    self.assertEquals(organization_show.call_count - count, expected_organization_show_calls)
AssertionError: 1 != 4

======================================================================
FAIL: test_list_datarequests_12 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 608, in test_list_datarequests
    self.assertEquals(expected_response['result'], response['result'])
AssertionError: Lists differ: [{'description': 'This is a ba... != [<MagicMock name='mock.dictize...

First differing element 0:
{'description': 'This is a basic description', 'organization_id': None, 'user': {'user': 3, 'id': 'user_id'}, 'id': 'example_uuidv4', 'close_time': None, 'user_id': 'example_uuidv4_user', 'open_time': '2020-06-10 18:41:45.498722', 'accepted_dataset': None, 'title': 'This is a title', 'accepted_dataset_id': None, 'followers': 3, 'closed': False, 'organization': None}
<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>

+ [<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>]
- [{'accepted_dataset': None,
-   'accepted_dataset_id': None,
-   'close_time': None,
-   'closed': False,
-   'description': 'This is a basic description',
-   'followers': 3,
-   'id': 'example_uuidv4',
-   'open_time': '2020-06-10 18:41:45.498722',
-   'organization': None,
-   'organization_id': None,
-   'title': 'This is a title',
-   'user': {'id': 'user_id', 'user': 3},
-   'user_id': 'example_uuidv4_user'}]

======================================================================
FAIL: test_list_datarequests_13 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 608, in test_list_datarequests
    self.assertEquals(expected_response['result'], response['result'])
AssertionError: Lists differ: [{'description': 'This is a ba... != [<MagicMock name='mock.dictize...

First differing element 0:
{'description': 'This is a basic description', 'organization_id': None, 'user': {'user': 3, 'id': 'user_id'}, 'id': 'example_uuidv4', 'close_time': None, 'user_id': 'example_uuidv4_user', 'open_time': '2020-06-10 18:41:45.498722', 'accepted_dataset': None, 'title': 'This is a title', 'accepted_dataset_id': None, 'followers': 3, 'closed': False, 'organization': None}
<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>

+ [<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>]
- [{'accepted_dataset': None,
-   'accepted_dataset_id': None,
-   'close_time': None,
-   'closed': False,
-   'description': 'This is a basic description',
-   'followers': 3,
-   'id': 'example_uuidv4',
-   'open_time': '2020-06-10 18:41:45.498722',
-   'organization': None,
-   'organization_id': None,
-   'title': 'This is a title',
-   'user': {'id': 'user_id', 'user': 3},
-   'user_id': 'example_uuidv4_user'}]

======================================================================
FAIL: test_list_datarequests_14 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 608, in test_list_datarequests
    self.assertEquals(expected_response['result'], response['result'])
AssertionError: Lists differ: [{'description': 'This is a ba... != [<MagicMock name='mock.dictize...

First differing element 0:
{'description': 'This is a basic description', 'organization_id': None, 'user': {'user': 3, 'id': 'user_id'}, 'id': 'example_uuidv4', 'close_time': None, 'user_id': 'example_uuidv4_user', 'open_time': '2020-06-10 18:41:45.498722', 'accepted_dataset': None, 'title': 'This is a title', 'accepted_dataset_id': None, 'followers': 3, 'closed': False, 'organization': None}
<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>

+ [<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>]
- [{'accepted_dataset': None,
-   'accepted_dataset_id': None,
-   'close_time': None,
-   'closed': False,
-   'description': 'This is a basic description',
-   'followers': 3,
-   'id': 'example_uuidv4',
-   'open_time': '2020-06-10 18:41:45.498722',
-   'organization': None,
-   'organization_id': None,
-   'title': 'This is a title',
-   'user': {'id': 'user_id', 'user': 3},
-   'user_id': 'example_uuidv4_user'}]

======================================================================
FAIL: test_list_datarequests_15 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 608, in test_list_datarequests
    self.assertEquals(expected_response['result'], response['result'])
AssertionError: Lists differ: [{'description': 'This is a ba... != [<MagicMock name='mock.dictize...

First differing element 0:
{'description': 'This is a basic description', 'organization_id': None, 'user': {'user': 3, 'id': 'user_id'}, 'id': 'example_uuidv4', 'close_time': None, 'user_id': 'example_uuidv4_user', 'open_time': '2020-06-10 18:41:45.498722', 'accepted_dataset': None, 'title': 'This is a title', 'accepted_dataset_id': None, 'followers': 3, 'closed': False, 'organization': None}
<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>

+ [<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>]
- [{'accepted_dataset': None,
-   'accepted_dataset_id': None,
-   'close_time': None,
-   'closed': False,
-   'description': 'This is a basic description',
-   'followers': 3,
-   'id': 'example_uuidv4',
-   'open_time': '2020-06-10 18:41:45.498722',
-   'organization': None,
-   'organization_id': None,
-   'title': 'This is a title',
-   'user': {'id': 'user_id', 'user': 3},
-   'user_id': 'example_uuidv4_user'}]

======================================================================
FAIL: test_list_datarequests_16 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 594, in test_list_datarequests
    self.assertEquals(organization_show.call_count - count, expected_organization_show_calls)
AssertionError: 1 != 4

======================================================================
FAIL: test_list_datarequests_2 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 608, in test_list_datarequests
    self.assertEquals(expected_response['result'], response['result'])
AssertionError: Lists differ: [{'description': 'This is a ba... != [<MagicMock name='mock.dictize...

First differing element 0:
{'description': 'This is a basic description', 'organization_id': None, 'user': {'user': 3, 'id': 'user_id'}, 'id': 'example_uuidv4', 'close_time': None, 'user_id': 'example_uuidv4_user', 'open_time': '2020-06-10 18:41:45.498722', 'accepted_dataset': None, 'title': 'This is a title', 'accepted_dataset_id': None, 'followers': 3, 'closed': False, 'organization': None}
<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>

+ [<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>]
- [{'accepted_dataset': None,
-   'accepted_dataset_id': None,
-   'close_time': None,
-   'closed': False,
-   'description': 'This is a basic description',
-   'followers': 3,
-   'id': 'example_uuidv4',
-   'open_time': '2020-06-10 18:41:45.498722',
-   'organization': None,
-   'organization_id': None,
-   'title': 'This is a title',
-   'user': {'id': 'user_id', 'user': 3},
-   'user_id': 'example_uuidv4_user'}]

======================================================================
FAIL: test_list_datarequests_3 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 608, in test_list_datarequests
    self.assertEquals(expected_response['result'], response['result'])
AssertionError: Lists differ: [{'description': 'This is a ba... != [<MagicMock name='mock.dictize...

First differing element 0:
{'description': 'This is a basic description', 'organization_id': None, 'user': {'user': 3, 'id': 'user_id'}, 'id': 'example_uuidv4', 'close_time': None, 'user_id': 'example_uuidv4_user', 'open_time': '2020-06-10 18:41:45.498722', 'accepted_dataset': None, 'title': 'This is a title', 'accepted_dataset_id': None, 'followers': 3, 'closed': False, 'organization': None}
<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>

+ [<MagicMock name='mock.dictize_datarequest()' id='139715371263184'>]
- [{'accepted_dataset': None,
-   'accepted_dataset_id': None,
-   'close_time': None,
-   'closed': False,
-   'description': 'This is a basic description',
-   'followers': 3,
-   'id': 'example_uuidv4',
-   'open_time': '2020-06-10 18:41:45.498722',
-   'organization': None,
-   'organization_id': None,
-   'title': 'This is a title',
-   'user': {'id': 'user_id', 'user': 3},
-   'user_id': 'example_uuidv4_user'}]

======================================================================
FAIL: test_list_datarequests_4 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 594, in test_list_datarequests
    self.assertEquals(organization_show.call_count - count, expected_organization_show_calls)
AssertionError: -3 != 3

======================================================================
FAIL: test_list_datarequests_5 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 594, in test_list_datarequests
    self.assertEquals(organization_show.call_count - count, expected_organization_show_calls)
AssertionError: -2 != 4

======================================================================
FAIL: test_list_datarequests_6 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 594, in test_list_datarequests
    self.assertEquals(organization_show.call_count - count, expected_organization_show_calls)
AssertionError: -3 != 3

======================================================================
FAIL: test_list_datarequests_7 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 594, in test_list_datarequests
    self.assertEquals(organization_show.call_count - count, expected_organization_show_calls)
AssertionError: -2 != 4

======================================================================
FAIL: test_list_datarequests_8 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 594, in test_list_datarequests
    self.assertEquals(organization_show.call_count - count, expected_organization_show_calls)
AssertionError: 0 != 3

======================================================================
FAIL: test_list_datarequests_9 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 594, in test_list_datarequests
    self.assertEquals(organization_show.call_count - count, expected_organization_show_calls)
AssertionError: 1 != 4

======================================================================
FAIL: test_show_datarequest_found_closed_0 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 439, in test_show_datarequest_found_closed
    self._test_show_datarequest_found(datarequest, org_checked, pkg_checked)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 412, in _test_show_datarequest_found
    self._check_basic_response(datarequest, result, default_user, org, pkg)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 76, in _check_basic_response
    self.assertEquals(datarequest.id, response['id'])
AssertionError: 'example_uuidv4' != <MagicMock name='mock.dictize_datarequest().__getitem__()' id='139715376109200'>

======================================================================
FAIL: test_show_datarequest_found_closed_1_org_id (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 439, in test_show_datarequest_found_closed
    self._test_show_datarequest_found(datarequest, org_checked, pkg_checked)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 412, in _test_show_datarequest_found
    self._check_basic_response(datarequest, result, default_user, org, pkg)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 76, in _check_basic_response
    self.assertEquals(datarequest.id, response['id'])
AssertionError: 'example_uuidv4' != <MagicMock name='mock.dictize_datarequest().__getitem__()' id='139715368429456'>

======================================================================
FAIL: test_show_datarequest_found_closed_2 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 439, in test_show_datarequest_found_closed
    self._test_show_datarequest_found(datarequest, org_checked, pkg_checked)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 412, in _test_show_datarequest_found
    self._check_basic_response(datarequest, result, default_user, org, pkg)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 76, in _check_basic_response
    self.assertEquals(datarequest.id, response['id'])
AssertionError: 'example_uuidv4' != <MagicMock name='mock.dictize_datarequest().__getitem__()' id='139715367615952'>

======================================================================
FAIL: test_show_datarequest_found_closed_3_org_id (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 439, in test_show_datarequest_found_closed
    self._test_show_datarequest_found(datarequest, org_checked, pkg_checked)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 412, in _test_show_datarequest_found
    self._check_basic_response(datarequest, result, default_user, org, pkg)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 76, in _check_basic_response
    self.assertEquals(datarequest.id, response['id'])
AssertionError: 'example_uuidv4' != <MagicMock name='mock.dictize_datarequest().__getitem__()' id='139715367820432'>

======================================================================
FAIL: test_show_datarequest_found_no_org_open (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckanext/datarequests/tests/test_actions.py", line 421, in test_show_datarequest_found_no_org_open
    self._test_show_datarequest_found(datarequest, False, False)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 412, in _test_show_datarequest_found
    self._check_basic_response(datarequest, result, default_user, org, pkg)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 76, in _check_basic_response
    self.assertEquals(datarequest.id, response['id'])
AssertionError: 'example_uuidv4' != <MagicMock name='mock.dictize_datarequest().__getitem__()' id='139715371871376'>

======================================================================
FAIL: test_show_datarequest_found_org_open (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckanext/datarequests/tests/test_actions.py", line 416, in test_show_datarequest_found_org_open
    self._test_show_datarequest_found(datarequest, True, False)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 412, in _test_show_datarequest_found
    self._check_basic_response(datarequest, result, default_user, org, pkg)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 76, in _check_basic_response
    self.assertEquals(datarequest.id, response['id'])
AssertionError: 'example_uuidv4' != <MagicMock name='mock.dictize_datarequest().__getitem__()' id='139715366563664'>

======================================================================
FAIL: test_update_datarequest_0 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 504, in test_update_datarequest
    self.assertEquals(test_data.update_request_data['title'], datarequest.title)
AssertionError: 'title' != 'titlea'

======================================================================
FAIL: test_update_datarequest_1 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 505, in test_update_datarequest
    self.assertEquals(test_data.update_request_data['description'], datarequest.description)
AssertionError: 'description' != 'This is a basic description'

======================================================================
FAIL: test_update_datarequest_2 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 505, in test_update_datarequest
    self.assertEquals(test_data.update_request_data['description'], datarequest.description)
AssertionError: 'description' != 'This is a basic description'

======================================================================
FAIL: test_update_datarequest_3 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 505, in test_update_datarequest
    self.assertEquals(test_data.update_request_data['description'], datarequest.description)
AssertionError: 'description' != 'This is a basic description'

======================================================================
FAIL: test_update_datarequest_4 (test_actions.ActionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_actions.py", line 505, in test_update_datarequest
    self.assertEquals(test_data.update_request_data['description'], datarequest.description)
AssertionError: 'description' != 'This is a basic description'

======================================================================
FAIL: test_initdb_not_initialized (test_db.DBTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckanext/datarequests/tests/test_db.py", line 165, in test_initdb_not_initialized
    model.meta.mapper.assert_any_call(db.DataRequest, table_data_request)
  File "/app/ckan/default/lib/python2.7/site-packages/mock/mock.py", line 1010, in assert_any_call
    ), cause)
  File "/app/ckan/default/lib/python2.7/site-packages/six.py", line 737, in raise_from
    raise value
AssertionError: mapper(<MagicMock spec='str' id='139715352372624'>, <MagicMock id='139715357698576'>) call not found

======================================================================
FAIL: test_close_0 (test_ui_controller.UIControllerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_ui_controller.py", line 747, in test_close
    self.assertEquals(expected_datasets, controller.c.datasets)
AssertionError: Lists differ: [{'name': 'pack1', 'title': 'p... != [{'name': None, 'title': 'pack...

First differing element 0:
{'name': 'pack1', 'title': 'pack1'}
{'name': None, 'title': 'pack1'}

- [{'name': 'pack1', 'title': 'pack1'}, {'name': 'pack2', 'title': 'pack2'}]
?           ^^^^^^^                              ^^^^^^^

+ [{'name': None, 'title': 'pack1'}, {'name': None, 'title': 'pack2'}]
?           ^^^^                              ^^^^


======================================================================
FAIL: test_close_1_organization_uuidv4 (test_ui_controller.UIControllerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_ui_controller.py", line 747, in test_close
    self.assertEquals(expected_datasets, controller.c.datasets)
AssertionError: Lists differ: [{'name': 'packo1', 'title': '... != [{'name': None, 'title': 'pack...

First differing element 0:
{'name': 'packo1', 'title': 'packo1'}
{'name': None, 'title': 'packo1'}

- [{'name': 'packo1', 'title': 'packo1'}, {'name': 'packo2', 'title': 'packo2'}]
?           ^^^^^ ^^                               ^^^^^ ^^

+ [{'name': None, 'title': 'packo1'}, {'name': None, 'title': 'packo2'}]
?           ^ ^^                               ^ ^^


======================================================================
FAIL: test_close_post_errors_0 (test_ui_controller.UIControllerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_ui_controller.py", line 786, in test_close_post_errors
    {'Accepted Dataset': 'error1, error2'}, close_datarequest)
  File "/app/ckanext/datarequests/tests/test_ui_controller.py", line 747, in test_close
    self.assertEquals(expected_datasets, controller.c.datasets)
AssertionError: Lists differ: [{'name': 'pack1', 'title': 'p... != [{'name': None, 'title': 'pack...

First differing element 0:
{'name': 'pack1', 'title': 'pack1'}
{'name': None, 'title': 'pack1'}

- [{'name': 'pack1', 'title': 'pack1'}, {'name': 'pack2', 'title': 'pack2'}]
?           ^^^^^^^                              ^^^^^^^

+ [{'name': None, 'title': 'pack1'}, {'name': None, 'title': 'pack2'}]
?           ^^^^                              ^^^^

-------------------- >> begin captured logging << --------------------
ckanext.datarequests.controllers.ui_controller: WARNING: {'Accepted Dataset': ['error1', 'error2']}
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: test_close_post_errors_1_organization_uuidv4 (test_ui_controller.UIControllerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckan/default/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/app/ckanext/datarequests/tests/test_ui_controller.py", line 786, in test_close_post_errors
    {'Accepted Dataset': 'error1, error2'}, close_datarequest)
  File "/app/ckanext/datarequests/tests/test_ui_controller.py", line 747, in test_close
    self.assertEquals(expected_datasets, controller.c.datasets)
AssertionError: Lists differ: [{'name': 'packo1', 'title': '... != [{'name': None, 'title': 'pack...

First differing element 0:
{'name': 'packo1', 'title': 'packo1'}
{'name': None, 'title': 'packo1'}

- [{'name': 'packo1', 'title': 'packo1'}, {'name': 'packo2', 'title': 'packo2'}]
?           ^^^^^ ^^                               ^^^^^ ^^

+ [{'name': None, 'title': 'packo1'}, {'name': None, 'title': 'packo2'}]
?           ^ ^^                               ^ ^^

-------------------- >> begin captured logging << --------------------
ckanext.datarequests.controllers.ui_controller: WARNING: {'Accepted Dataset': ['error1', 'error2']}
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: test_close_invalid_accepted_dataset (test_validator.ValidatorTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/ckanext/datarequests/tests/test_validator.py", line 121, in test_close_invalid_accepted_dataset
    validator.tk.get_validator.assert_called_once_with('package_name_exists')
  File "/app/ckan/default/lib/python2.7/site-packages/mock/mock.py", line 956, in assert_called_once_with
    raise AssertionError(msg)
AssertionError: Expected 'get_validator' to be called once. Called 0 times.

----------------------------------------------------------------------
Ran 282 tests in 2.478s

FAILED (errors=1, failures=43)
```

MarkCalvert replied:
Have fixed the unit tests and re-enabled on them circleci

Commit:
[DQL2-33]
- Updated approx_publishing_date to use 'render_datetime 'helper function timestamp string as a localised date

duttonw commented:
latest commit 8c8ec0c https://app.circleci.com/pipelines/github/MarkCalvert/ckanext-datarequests/4/workflows/c01a2b24-a188-4d8c-b481-911f742dadc1/jobs/4/steps

Ran 282 tests in 2.446s 
FAILED (errors=1, failures=43)
 
cucumber tests
4 features passed, 0 failed, 0 skipped
50 scenarios passed, 0 failed, 0 skipped
280 steps passed, 0 failed, 0 skipped, 0 undefined
Took 5m0.147s
 
why does unit tests fail?

MarkCalvert commented:
I have not investigated yet, but they could be failing from the new foreign key constraints I added to the datarequest table for organisation_id and accepted_id.
This change was made to utilise the sqlalchemy relationships to load the relationship table data instead of manually loading these relationships with the package_show and organisation_show get_action calls.
So if the unit tests are using bogus data for the organisation and accepted datasets, then those tests will probably fail.
The unit tests were turned off in circleci because all the BDD tests passed, and I  did not want this to be a blocker for QA for testing. I intend to investigate these unit tests issues later on or in the Data Requests - Open Source Considerations story (https://salsadigital.atlassian.net/browse/DQL2-48)
FYI we have not added/updated any of this extensions unit tests so some discovery time will be required to understand how the original developer set them up.

MarkCalvert commented:
Another option, which may be quicker is to revert those foreign key constraints and relationships rather than trying to investigate all of the unit tests and update them with correct data. I will investigate to see if they are actually the issue or if its something unrelated.

Commit:
[DQL2-31]
- Reverted changes back to original way of using '_dictize_datarequest' & '_undictize_datarequest_basic' and not using model relationships
- Re-enabled unit test failures
- Updated unit tests to default to using 'tk.h.closing_circumstances_enabled = False'

Commit:
[DQL2-31]
Increase random number range to descrease change of duplicates

Commit:
[DQL-31]
- Updated CLOSE_CIRCUMSTANCE_MAX_LENGTH from 100 to 255

Commit:
[DQL-31]
- Updated variable name data_request to 'datarequest' to keep same standard in the file
- Updated _undictize_datarequest_closing_circumstances default values to None when empty string
